### PR TITLE
Merge multiple Cobertura reports.

### PR DESCRIPTION
**Pull request by tren-parasoft**

### CODE REVIEW

File: PublishParasoftResults/CoverageReportMerger.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    mergeCoberturaReports = (reportPath: string[]): string => {
+        if(reportPath.length == 1) {
+            return reportPath[0];
+        }
+        let path = reportPath[0];
+        const baseReport = this.processToJson(reportPath[0]);
+        for(let i = 1; i <reportPath.length; i++) {
+            const report: CoberturaReport = this.processToJson(reportPath[i]);
+            // Todo: Make sure these two reports have same files and file contents are same then do the merge
```
[the remark this answers is not in the document]
Comment (yshi-parasoft) [reply PublishParasoftResults/CoverageReportMerger.ts:47]: I can see that new `Package`, `Class` and `Line` are not handled in implementation of merging.
Are we going to handle them in another task?

File: PublishParasoftResults/CoverageReportMerger.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            this.mergeCoberturaReport(baseReport, report);
+            path = `${this.defaultWorkingDirectory}/parasoft-merged-cobertura.xml`;
+        }
+        fs.writeFileSync(path, this.processToXML(baseReport), 'utf-8');
+        return path;
+    };
+
+    private processToJson = (reportPath: string): CoberturaReport => {
```
[the remark this answers is not in the document]
Comment (tren-parasoft) [reply PublishParasoftResults/CoverageReportMerger.ts:55]: xml2js will trans <classes> and <lines> to array, but Map is better choice here.

File: PublishParasoftResults/CoverageReportMerger.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                const lineRate = <string>node.attributes['line-rate'];
+
+                if (!fileName || !name) {
+                    throw new Error("error in Cobertura code coverage report: failed to parse 'filename/name' attribute.");
+                } if (!lineRate || isNaN(parseFloat(lineRate))) {
+                    throw new Error("error in Cobertura code coverage report: failed to parse 'line-rate' attribute.");
+                } else {
+                    classFile.name = name;
+                    classFile.fileName = fileName;
+                    classFile.classId = `${name}-${fileName}`;
```
[the remark this answers is not in the document]
Comment (tren-parasoft) [reply PublishParasoftResults/CoverageReportMerger.ts:122]: Adding **classId** here is for the map Key,
image
In this report , these two file have same **fileName**, but their **name** are different and the **name** also could be duplicated.

File: PublishParasoftResults/ParaReportPublishService.ts [hunk 4]
```diff
@@ -581,7 +582,8 @@ export class ParaReportPublishService {
     private async processCoberturaResults(): Promise<void> {
         if (this.coberturaReports.length > 0) {
             const tempFolder = path.join(this.getTempFolder(), 'CodeCoverageHtml');
-            const coverageReport: string = <string> this.coberturaReports[this.coberturaReports.length - 1];
+            const merger = new CoverageReportService(this.defaultWorkingDirectory);
```
Comment (rwang-parasoft) [inline PublishParasoftResults/ParaReportPublishService.ts:585]: Rename merger to coverageReportService.

File: PublishParasoftResults/CoverageReportService.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    lineNumber: number;
+    lineHash: string;
+    hits: number;
+}
+
+export class CoverageReportService {
+    private readonly destFolderPath: string;
+
+    constructor(defaultWorkingDirectory: string) {
+        this.destFolderPath = defaultWorkingDirectory;
```
Comment (rwang-parasoft) [inline PublishParasoftResults/CoverageReportService.ts:51]: The value of defaultWorkingDirectory can be found by tl.getVariable('System.DefaultWorkingDirectory'), so it is not necessary to pass the value through constructor parameter.

File: PublishParasoftResults/CoverageReportService.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    name: string;
+    lineRate: number;
+    classes: Map<string, CoberturaClass>;
+}
+
+type CoberturaClass = {
+    fileName: string;
+    name: string;
+    lineRate: number;
+    classId: string;
```
Comment (yshi-parasoft) [inline PublishParasoftResults/CoverageReportService.ts:39]: Please add comment here to indicate it's "name + filename"

File: PublishParasoftResults/CoverageReportService.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        if(reportPaths.length == 1) {
+            fs.copyFileSync(reportPaths[0], this.mergedCoberturaReportPath);
+            return this.mergedCoberturaReportPath;
+        }
+
+        let baseCoverage = this.processXMLToObj(reportPaths[0]);
+        for(let i = 1; i < reportPaths.length; i++) {
+            const reportToMerge: CoberturaCoverage = this.processXMLToObj(reportPaths[i]);
+            try {
+                tl.debug(`Merging cobertura reort: ${reportPaths[i]}`);
```
Comment (yshi-parasoft) [inline PublishParasoftResults/CoverageReportService.ts:66]: Change all "cobertura" in messages to "Cobertura"

File: PublishParasoftResults/CoverageReportService.ts [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+                baseCoverage = this.mergeCoberturaCoverage(lodash.cloneDeep(baseCoverage), reportToMerge);
+            } catch (error) {
+                if (error instanceof Error) {
+                    tl.warning(`Skip merging cobertura reort: ${reportPaths[i]} to ${this.mergedCoberturaReportPath}, due to ${error.message}`);
+                } else {
+                    tl.warning(`Follow cobertura reports are conflict: ${reportPaths[i]}, ${this.mergedCoberturaReportPath}`);
+                }
+            }
+        }
+
```
Comment (yshi-parasoft) [inline PublishParasoftResults/CoverageReportService.ts:76]: Please add a comment here to explain what `updateAttributes `does.

File: PublishParasoftResults/CoverageReportService.ts [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+            return this.mergedCoberturaReportPath;
+        }
+
+        let baseCoverage = this.processXMLToObj(reportPaths[0]);
+        for(let i = 1; i < reportPaths.length; i++) {
+            const reportToMerge: CoberturaCoverage = this.processXMLToObj(reportPaths[i]);
+            try {
+                tl.debug(`Merging cobertura reort: ${reportPaths[i]}`);
+                baseCoverage = this.mergeCoberturaCoverage(lodash.cloneDeep(baseCoverage), reportToMerge);
+            } catch (error) {
```
Comment (yshi-parasoft) [inline PublishParasoftResults/CoverageReportService.ts:68]: It appears that both Saxon parsing and conflicts during report merging can cause an `Error`. 
Therefore, a generic error message for that is fine. Let's change it to: `Skipped merging Cobertura report "xxx": ${error.message}`
Why do we consider the `else` case as a conflicting case?

**Pull request merged**